File: 2015/04_The_Ideal_Stocking_Stuffer/Day04.clj
; Day 4: The Ideal Stocking Stuffer
; https://adventofcode.com/2015/day/4

(ns day04
  (:require [clojure.string :as str])
  (:import [java.security MessageDigest]))

(defn md5 [input]
  (let [md (MessageDigest/getInstance "MD5")
        bytes (.digest md (.getBytes input))]
    (apply str (map #(format "%02x" %) bytes))))

(defn part1 [input]
  (loop [running-number 0]
    (if (.startsWith (md5 (str input running-number)) "00000")
      running-number
      (recur (inc running-number)))))

(defn part2 [input]
  (loop [running-number 0]
    (if (.startsWith (md5 (str input running-number)) "000000")
      running-number
      (recur (inc running-number)))))

(def input1 "abcdef")
(def input2 "pqrstuv")
(def input-test "bgvyzdsv")

(println "Part 1:" (part1 input1)) ; 609043
(println "Part 1:" (part1 input2)) ; 1048970
(println "Part 1:" (part1 input-test)) ; 254575

(println "Part 2:" (part2 input1)) ; 6742839
(println "Part 2:" (part2 input2)) ; 5714438
(println "Part 2:" (part2 input-test)) ; 1038736

File: 2015/04_The_Ideal_Stocking_Stuffer/Day04.kt
// Day 4: The Ideal Stocking Stuffer
// https://adventofcode.com/2015/day/4

import java.security.MessageDigest

fun md5(input: String): String {
    val bytes = MessageDigest.getInstance("MD5").digest(input.toByteArray())
    return bytes.joinToString("") { "%02x".format(it) }
}

fun part1(input: String): Int {
    var runningNumber = 0
    while (!md5("$input$runningNumber").startsWith("00000")) {
        runningNumber++
    }
    return runningNumber
}

fun part2(input: String): Int {
    var runningNumber = 0
    while (!md5("$input$runningNumber").startsWith("000000")) {
        runningNumber++
    }
    return runningNumber
}

fun main() {
    val input1 = "abcdef"
    val input2 = "pqrstuv"
    val inputTest = "bgvyzdsv"

    println("Part 1: ${part1(input1)}") // 609043
    println("Part 1: ${part1(input2)}") // 1048970
    println("Part 1: ${part1(inputTest)}") // 254575

    println("Part 2: ${part2(input1)}") // 6742839
    println("Part 2: ${part2(input2)}") // 5714438
    println("Part 2: ${part2(inputTest)}") // 1038736
}